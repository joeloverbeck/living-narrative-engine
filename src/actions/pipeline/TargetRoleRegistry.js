/**
 * @file TargetRoleRegistry.js
 * @description Provides canonical role definitions and helper utilities for target validation.
 */

export const LEGACY_TARGET_ROLE = 'target';
export const PRIMARY_ROLE = 'primary';
export const SECONDARY_ROLE = 'secondary';
export const TERTIARY_ROLE = 'tertiary';
export const ACTOR_ROLE = 'actor';

export const ALL_MULTI_TARGET_ROLES = Object.freeze([
  PRIMARY_ROLE,
  SECONDARY_ROLE,
  TERTIARY_ROLE,
]);

export const LEGACY_ROLES = Object.freeze([LEGACY_TARGET_ROLE]);
export const ALL_TARGET_ROLES = Object.freeze([
  LEGACY_TARGET_ROLE,
  ...ALL_MULTI_TARGET_ROLES,
]);

export const ALL_ROLES_WITH_ACTOR = Object.freeze([
  ACTOR_ROLE,
  ...ALL_MULTI_TARGET_ROLES,
]);

/**
 * Determines whether the provided role is supported by the registry.
 *
 * @param {string} role - Target role to validate.
 * @returns {boolean} True when the role is recognized.
 */
export function isSupportedRole(role) {
  return ALL_TARGET_ROLES.includes(role) || role === ACTOR_ROLE;
}

/**
 * Determine the placeholder definition for a role.
 *
 * @param {string} role - Role to inspect.
 * @param {object|undefined|null} placeholderSource - Placeholder metadata collection.
 * @returns {string|undefined} Placeholder identifier when available.
 */
export function getPlaceholderForRole(role, placeholderSource) {
  if (!placeholderSource || typeof placeholderSource !== 'object') {
    return undefined;
  }

  const entry = placeholderSource[role];
  if (!entry || typeof entry !== 'object') {
    return undefined;
  }

  return entry.placeholder;
}

/**
 * Extract roles that define required components.
 *
 * @param {object|undefined|null} requirements - Required components definition.
 * @returns {string[]} Ordered list of roles with requirements.
 */
export function getRolesWithRequirements(requirements) {
  if (!requirements || typeof requirements !== 'object') {
    return [];
  }

  const roles = [];
  if (Array.isArray(requirements[LEGACY_TARGET_ROLE]) && requirements[LEGACY_TARGET_ROLE].length > 0) {
    roles.push(LEGACY_TARGET_ROLE);
  }

  for (const role of ALL_MULTI_TARGET_ROLES) {
    if (Array.isArray(requirements[role]) && requirements[role].length > 0) {
      roles.push(role);
    }
  }

  return roles;
}

/**
 * Determine if an action definition represents the legacy target payload.
 *
 * @param {import('../../interfaces/IGameDataRepository.js').ActionDefinition|undefined|null} actionDef - Action definition to inspect.
 * @returns {boolean} True when the action follows the legacy single-target format.
 */
export function isLegacyTargetPayload(actionDef) {
  if (!actionDef) {
    return false;
  }

  if (actionDef.target_entity) {
    return true;
  }

  if (actionDef.resolvedTargets && actionDef.resolvedTargets[LEGACY_TARGET_ROLE]) {
    return true;
  }

  if (actionDef.targets && actionDef.targets[LEGACY_TARGET_ROLE]) {
    return true;
  }

  return false;
}

/**
 * Determine if an action definition represents the multi-target payload.
 *
 * @param {import('../../interfaces/IGameDataRepository.js').ActionDefinition|undefined|null} actionDef - Action definition to inspect.
 * @returns {boolean} True when the action exposes multi-target roles.
 */
export function isMultiTargetPayload(actionDef) {
  if (!actionDef) {
    return false;
  }

  if (actionDef.target_entities) {
    return ALL_MULTI_TARGET_ROLES.some((role) => actionDef.target_entities[role]);
  }

  if (actionDef.resolvedTargets) {
    return ALL_MULTI_TARGET_ROLES.some((role) => actionDef.resolvedTargets[role]);
  }

  if (actionDef.targetDefinitions) {
    return ALL_MULTI_TARGET_ROLES.some((role) => actionDef.targetDefinitions[role]);
  }

  return false;
}

export default {
  LEGACY_TARGET_ROLE,
  PRIMARY_ROLE,
  SECONDARY_ROLE,
  TERTIARY_ROLE,
  ACTOR_ROLE,
  ALL_MULTI_TARGET_ROLES,
  ALL_TARGET_ROLES,
  ALL_ROLES_WITH_ACTOR,
  LEGACY_ROLES,
  isSupportedRole,
  getPlaceholderForRole,
  getRolesWithRequirements,
  isLegacyTargetPayload,
  isMultiTargetPayload,
};
