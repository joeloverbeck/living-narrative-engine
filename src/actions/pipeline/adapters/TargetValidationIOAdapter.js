/**
 * @file TargetValidationIOAdapter.js
 * @description Normalizes pipeline context for the target validation stage and
 * provides utilities to rebuild the original structure after validation.
 */

import {
  ACTOR_ROLE,
  ALL_MULTI_TARGET_ROLES,
  LEGACY_TARGET_ROLE,
} from '../TargetRoleRegistry.js';
import { deepClone } from '../../../utils/cloneUtils.js';

/**
 * @typedef {import('../../../entities/entity.js').default} Entity
 */

/**
 * @typedef {import('../../../interfaces/IGameDataRepository.js').ActionDefinition} ActionDefinition
 */

/**
 * @typedef {object} TargetValidationItem
 * @property {ActionDefinition} actionDef - Action definition under validation.
 * @property {object|null} resolvedTargets - Canonical resolved targets map including actor when available.
 * @property {object|null} targetDefinitions - Target definition metadata for placeholder lookups.
 * @property {Array<object>} targetContexts - Target context entries associated with the action.
 * @property {object|null} placeholderSource - Source collection for placeholder metadata.
 * @property {'actionsWithTargets'|'candidateActions'} sourceFormat - Source collection identifier.
 * @property {number} originalIndex - Index of the action in the original collection.
 * @property {object|null} originalRef - Reference to the original `actionsWithTargets` entry when applicable.
 */

/**
 * @typedef {object} TargetValidationMetadata
 * @property {Entity|null} actor - The actor entity provided to the stage.
 * @property {object|null} sharedResolvedTargetsRef - Shared resolved targets reference from the context.
 */

/**
 * @typedef {object} NormalizedTargetValidation
 * @property {'actionsWithTargets'|'candidateActions'|'empty'} format - Source format identifier.
 * @property {TargetValidationItem[]} items - Canonical item collection.
 * @property {TargetValidationMetadata} metadata - Shared metadata required for rebuild.
 */

/**
 * @typedef {object} RebuildParams
 * @property {'actionsWithTargets'|'candidateActions'|'empty'} format - Original format identifier.
 * @property {TargetValidationItem[]} items - Original normalized items (ordered).
 * @property {TargetValidationMetadata} metadata - Shared metadata captured during normalization.
 * @property {TargetValidationItem[]} validatedItems - Items that passed validation.
 */

/**
 * @typedef {object} RebuildResult
 * @property {{actionsWithTargets?: object[], candidateActions?: ActionDefinition[]}} data - Restored payload.
 * @property {boolean} continueProcessing - Indicates if downstream stages should continue processing.
 */

/**
 * IO adapter that converts pipeline context into a canonical representation for
 * the target validation stage.
 */
export class TargetValidationIOAdapter {
  /**
   * Normalize the pipeline context into canonical validation items.
   *
   * @param {object} pipelineContext - Raw pipeline context passed to the stage.
   * @returns {NormalizedTargetValidation} Canonical representation for validation.
   */
  normalize(pipelineContext) {
    const actionsWithTargets = Array.isArray(pipelineContext?.actionsWithTargets)
      ? pipelineContext.actionsWithTargets
      : [];
    const candidateActions = Array.isArray(pipelineContext?.candidateActions)
      ? pipelineContext.candidateActions
      : [];

    /** @type {'actionsWithTargets'|'candidateActions'|'empty'} */
    let format = 'empty';
    let sources = [];

    if (actionsWithTargets.length > 0) {
      format = 'actionsWithTargets';
      sources = actionsWithTargets;
    } else if (candidateActions.length > 0) {
      format = 'candidateActions';
      sources = candidateActions;
    }

    const actor = pipelineContext?.actor ?? null;
    const sharedResolvedTargetsRef = pipelineContext?.resolvedTargets ?? null;

    /** @type {TargetValidationItem[]} */
    const items = sources.map((source, index) => {
      const { actionDef } = this.#extractActionDefinition(source, format);
      const placeholderSource =
        this.#extractPlaceholderSource(source, actionDef) ?? null;
      const targetDefinitions = placeholderSource ?? null;

      const resolvedTargets = this.#buildResolvedTargets({
        source,
        actionDef,
        actor,
        format,
        sharedResolvedTargetsRef,
      });

      const targetContexts = this.#extractTargetContexts(source);

      return {
        actionDef,
        resolvedTargets,
        targetDefinitions,
        targetContexts,
        placeholderSource,
        sourceFormat: format === 'empty' ? 'candidateActions' : format,
        originalIndex: index,
        originalRef: format === 'actionsWithTargets' ? source : null,
      };
    });

    return {
      format,
      items,
      metadata: {
        actor,
        sharedResolvedTargetsRef,
      },
    };
  }

  /**
   * Rebuild the original payload structure from validated items.
   *
   * @param {RebuildParams} params - Parameters used to rebuild the payload.
   * @returns {RebuildResult} Restored data payload and continuation flag.
   */
  rebuild({ format, items, metadata, validatedItems }) {
    if (format === 'empty') {
      return {
        data: { candidateActions: [] },
        continueProcessing: false,
      };
    }

    const validatedSet = new Set(validatedItems);
    const orderedValidatedItems = items.filter((item) => validatedSet.has(item));
    const continueProcessing = orderedValidatedItems.length > 0;

    if (format === 'actionsWithTargets') {
      const rebuiltActions = orderedValidatedItems.map((item) => {
        const base = item.originalRef ? { ...item.originalRef } : { actionDef: item.actionDef };
        const sanitizedTargets = this.#sanitizeResolvedTargets(item.resolvedTargets);

        return {
          ...base,
          actionDef: item.actionDef,
          resolvedTargets: sanitizedTargets ? deepClone(sanitizedTargets) : sanitizedTargets,
          targetDefinitions: item.targetDefinitions ?? base.targetDefinitions ?? null,
          targetContexts: Array.isArray(item.targetContexts)
            ? [...item.targetContexts]
            : base.targetContexts,
        };
      });

      return {
        data: { actionsWithTargets: rebuiltActions },
        continueProcessing,
      };
    }

    // Legacy format rebuild
    const rebuiltCandidateActions = orderedValidatedItems.map((item) => item.actionDef);

    return {
      data: { candidateActions: rebuiltCandidateActions },
      continueProcessing,
    };
  }

  /**
   * Extract the action definition from the source item.
   *
   * @param {object} source - Source entry from pipeline context.
   * @param {'actionsWithTargets'|'candidateActions'|'empty'} format - Input format.
   * @returns {{ actionDef: ActionDefinition }} Extracted action definition.
   */
  #extractActionDefinition(source, format) {
    if (format === 'actionsWithTargets' && source && source.actionDef) {
      return { actionDef: source.actionDef };
    }

    return { actionDef: source };
  }

  /**
   * Extract placeholder source metadata from the source entry or action definition.
   *
   * @param {object} source - Source entry from pipeline context.
   * @param {ActionDefinition} actionDef - Action definition under normalization.
   * @returns {object|undefined} Placeholder source metadata.
   */
  #extractPlaceholderSource(source, actionDef) {
    if (source?.targetDefinitions) {
      return source.targetDefinitions;
    }

    if (actionDef?.targetDefinitions) {
      return actionDef.targetDefinitions;
    }

    if (actionDef?.targets) {
      return actionDef.targets;
    }

    return undefined;
  }

  /**
   * Build resolved targets map for the action.
   *
   * @param {object} params - Parameters required to build resolved targets.
   * @param {object} params.source - Original source entry.
   * @param {ActionDefinition} params.actionDef - Action definition being normalized.
   * @param {Entity|null} params.actor - Actor entity for the stage.
   * @param {'actionsWithTargets'|'candidateActions'|'empty'} params.format - Source format identifier.
   * @param {object|null} params.sharedResolvedTargetsRef - Shared context resolved targets reference.
   * @returns {object|null} Canonical resolved targets including actor when available.
   */
  #buildResolvedTargets({
    source,
    actionDef,
    actor,
    format,
    sharedResolvedTargetsRef,
  }) {
    const resolvedFromSource =
      (format === 'actionsWithTargets' ? source?.resolvedTargets : null) ??
      actionDef?.resolvedTargets ??
      sharedResolvedTargetsRef ??
      null;

    const cloned = resolvedFromSource ? deepClone(resolvedFromSource) : null;

    if (cloned) {
      if (actor) {
        cloned[ACTOR_ROLE] = actor;
      }
      return cloned;
    }

    if (actionDef?.target_entity) {
      const targets = { [LEGACY_TARGET_ROLE]: deepClone(actionDef.target_entity) };
      if (actor) {
        targets[ACTOR_ROLE] = actor;
      }
      return targets;
    }

    if (actionDef?.target_entities) {
      const targets = {};
      for (const role of ALL_MULTI_TARGET_ROLES) {
        if (actionDef.target_entities[role]) {
          targets[role] = deepClone(actionDef.target_entities[role]);
        }
      }
      if (actor) {
        targets[ACTOR_ROLE] = actor;
      }
      return Object.keys(targets).length > 0 ? targets : actor ? { [ACTOR_ROLE]: actor } : null;
    }

    if (actor) {
      return { [ACTOR_ROLE]: actor };
    }

    return null;
  }

  /**
   * Extract target contexts from the source entry when available.
   *
   * @param {object} source - Source entry from pipeline context.
   * @returns {Array<object>} Target context array.
   */
  #extractTargetContexts(source) {
    if (Array.isArray(source?.targetContexts)) {
      return source.targetContexts.map((contextEntry) => deepClone(contextEntry));
    }

    return [];
  }

  /**
   * Remove actor role from resolved targets when persisting back to action definitions.
   *
   * @param {object|null} resolvedTargets - Resolved targets including actor.
   * @returns {object|null} Sanitized resolved targets without actor entry.
   */
  #sanitizeResolvedTargets(resolvedTargets) {
    if (!resolvedTargets || typeof resolvedTargets !== 'object') {
      return null;
    }

    const sanitized = {};
    for (const [role, value] of Object.entries(resolvedTargets)) {
      if (role === ACTOR_ROLE) {
        continue;
      }
      sanitized[role] = value;
    }

    return Object.keys(sanitized).length > 0 ? sanitized : null;
  }
}

export default TargetValidationIOAdapter;
