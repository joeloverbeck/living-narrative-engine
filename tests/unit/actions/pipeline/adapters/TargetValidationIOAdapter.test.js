import TargetValidationIOAdapter from '../../../../../src/actions/pipeline/adapters/TargetValidationIOAdapter.js';
import { LEGACY_TARGET_ROLE, PRIMARY_ROLE } from '../../../../../src/actions/pipeline/TargetRoleRegistry.js';

describe('TargetValidationIOAdapter', () => {
  const actor = { id: 'actor-1' };

  it('normalizes and rebuilds actionsWithTargets payloads', () => {
    const context = {
      actor,
      actionsWithTargets: [
        {
          actionDef: {
            id: 'multi-1',
            required_components: {},
          },
          resolvedTargets: {
            [PRIMARY_ROLE]: [{ id: 'entity-1' }],
          },
          targetDefinitions: {
            [PRIMARY_ROLE]: { placeholder: 'target.primary' },
          },
          targetContexts: [
            {
              type: 'entity',
              placeholder: 'target.primary',
              entityId: 'entity-1',
            },
          ],
        },
      ],
    };

    const adapter = new TargetValidationIOAdapter();
    const normalized = adapter.normalize(context);

    expect(normalized).toMatchSnapshot();

    const rebuilt = adapter.rebuild({
      format: normalized.format,
      items: normalized.items,
      metadata: normalized.metadata,
      validatedItems: normalized.items,
    });

    expect(rebuilt.continueProcessing).toBe(true);
    expect(rebuilt.data.actionsWithTargets).toEqual(context.actionsWithTargets);
  });

  it('rebuilds candidateActions payloads with sanitized targets', () => {
    const context = {
      actor,
      resolvedTargets: {
        [LEGACY_TARGET_ROLE]: [{ id: 'legacy-1' }],
      },
      candidateActions: [
        {
          id: 'legacy-action',
          resolvedTargets: {
            [LEGACY_TARGET_ROLE]: [{ id: 'legacy-1' }],
          },
          required_components: {},
          targetDefinitions: {
            [LEGACY_TARGET_ROLE]: { placeholder: 'target.single' },
          },
        },
        {
          id: 'secondary-action',
          target_entities: {
            [PRIMARY_ROLE]: { id: 'entity-2' },
          },
          required_components: {},
        },
      ],
    };

    const adapter = new TargetValidationIOAdapter();
    const normalized = adapter.normalize(context);

    expect(normalized.format).toBe('candidateActions');
    expect(normalized.items).toHaveLength(2);

    const validatedItems = [normalized.items[0]];
    validatedItems[0].resolvedTargets = {
      ...validatedItems[0].resolvedTargets,
      [LEGACY_TARGET_ROLE]: [{ id: 'legacy-1' }],
    };

    const rebuilt = adapter.rebuild({
      format: normalized.format,
      items: normalized.items,
      metadata: normalized.metadata,
      validatedItems,
    });

    expect(rebuilt.continueProcessing).toBe(true);
    expect(rebuilt.data.candidateActions).toHaveLength(1);
    expect(rebuilt.data.candidateActions[0].resolvedTargets).toEqual({
      [LEGACY_TARGET_ROLE]: [{ id: 'legacy-1' }],
    });
    expect(context.resolvedTargets[LEGACY_TARGET_ROLE]).toEqual([
      { id: 'legacy-1' },
    ]);
  });
});
