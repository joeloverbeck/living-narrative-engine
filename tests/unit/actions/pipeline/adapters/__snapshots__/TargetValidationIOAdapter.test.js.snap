// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TargetValidationIOAdapter normalizes and rebuilds actionsWithTargets payloads 1`] = `
{
  "format": "actionsWithTargets",
  "items": [
    {
      "actionDef": {
        "id": "multi-1",
        "required_components": {},
      },
      "originalIndex": 0,
      "originalRef": {
        "actionDef": {
          "id": "multi-1",
          "required_components": {},
        },
        "resolvedTargets": {
          "primary": [
            {
              "id": "entity-1",
            },
          ],
        },
        "targetContexts": [
          {
            "entityId": "entity-1",
            "placeholder": "target.primary",
            "type": "entity",
          },
        ],
        "targetDefinitions": {
          "primary": {
            "placeholder": "target.primary",
          },
        },
      },
      "placeholderSource": {
        "primary": {
          "placeholder": "target.primary",
        },
      },
      "resolvedTargets": {
        "actor": {
          "id": "actor-1",
        },
        "primary": [
          {
            "id": "entity-1",
          },
        ],
      },
      "sourceFormat": "actionsWithTargets",
      "targetContexts": [
        {
          "entityId": "entity-1",
          "placeholder": "target.primary",
          "type": "entity",
        },
      ],
      "targetDefinitions": {
        "primary": {
          "placeholder": "target.primary",
        },
      },
    },
  ],
  "metadata": {
    "actor": {
      "id": "actor-1",
    },
    "sharedResolvedTargetsRef": null,
  },
}
`;
